fe/fe-core/src/main/java/com/starrocks/http/rest/TransactionLoadAction.java

```diff
@@ -184,192 +213,593 @@ public void executeWithoutPassword(BaseRequest request, BaseResponse response) t
+                    }
+                    return null;
             }
+
+            return result;
         }
 
-        if (channelIdStr == null) {
-            // 2. redirect transaction op to BE
-            synchronized (this) {
-                // 2.1 save label->be map when begin transaction, so that subsequent operator can send to same BE
-                if (op.equalsIgnoreCase(TXN_BEGIN)) {
-                    nodeID = GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo().seqChooseBackendOrComputeId();
+        private Long getNodeId(Operation op, String label) throws UserException {
+            Long nodeId;
+            synchronized (acquireNodeIdLock) {
+                // save label->be map when begin transaction, so that subsequent operator can send to same BE
+                if (TXN_BEGIN.equals(op)) {
+                    nodeId = GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo().seqChooseBackendOrComputeId();
                     // txnNodeMap is LRU cache, it atomic remove unused entry
-                    txnNodeMap.put(label, nodeID);
-                } else if (channelIdStr == null) {
-                    nodeID = txnNodeMap.get(label);
+                    txnNodeMap.put(label, nodeId);
+                } else {
+                    nodeId = txnNodeMap.get(label);
                 }
             }
+
+            if (nodeId == null) {
+                throw new UserException(String.format(
+                        "Transaction with op[%s] and label[%s] has no node.", op.getValue(), label));
+            }
+
+            return nodeId;
         }
 
-        if (op.equalsIgnoreCase(TXN_BEGIN) && channelIdStr != null) {
-            TransactionResult resp = new TransactionResult();
-            long timeoutMillis = 20000;
-            if (timeout != null) {
-                timeoutMillis = Long.parseLong(timeout) * 1000;
+    }
+
+    /* helper classes */
+
+    private static class OpMetrics {
+        LongCounterMetric opRunningNum;
+        Histogram opLatencyMs;
+
+        static OpMetrics of(LongCounterMetric opRunningNum, Histogram opLatencyMs) {
+            OpMetrics metrics = new OpMetrics();
+            metrics.opRunningNum = opRunningNum;
+            metrics.opLatencyMs = opLatencyMs;
+            return metrics;
+        }
+    }
+
+    /**
+     * Request params parser, validator and holder.
+     */
+    private static class RequestParams {
+
+        /* headers */
+
+        private final String dbName;
+        private final String tableName;
+        private final String label;
+        private final Operation operation;
+        private final Long timeoutMillis;
+        private final Channel channel;
+
+        /* queries */
+
+        private final LoadJobSourceType sourceType;
+
+        /* body */
+
+        private final Body body;
+
+        public RequestParams(String dbName,
```

Comment:
```suggestion
        private RequestParams(String dbName,
```

Reply:
`RequestParams` is already a private class.